src/node_kernels/MomentumMassBDFNodeKernel.C

```diff
@@ -0,0 +1,93 @@
+#include "utils/StkHelpers.h"
+
+namespace sierra{
+namespace nalu{
+
+MomentumMassBDFNodeKernel::MomentumMassBDFNodeKernel(
+  const stk::mesh::BulkData& bulk,
+  ScalarFieldType *scalarQ
+) : NGPNodeKernel<MomentumMassBDFNodeKernel>(),
+    nDim_(bulk.mesh_meta_data().spatial_dimension())
+{
+  const auto& meta = bulk.mesh_meta_data();
+
+  velocityNID_ = scalarQ->field_of_state(stk::mesh::StateN).mesh_meta_data_ordinal();
+
+  if (scalarQ->number_of_states() == 2)
+    velocityNm1ID_ = velocityNID_;
+  else
+    velocityNm1ID_ = scalarQ->field_of_state(stk::mesh::StateNM1).mesh_meta_data_ordinal();
+
+  velocityNp1ID_ = scalarQ->field_of_state(stk::mesh::StateNP1).mesh_meta_data_ordinal();
+
+  densityNID_ = get_field_ordinal(meta, "density", stk::mesh::StateN);
+
+  if (scalarQ->number_of_states() == 2)
+    densityNm1ID_ = densityNID_;
+  else
+    densityNm1ID_ = get_field_ordinal(meta, "density", stk::mesh::StateNM1);
+
+  densityNp1ID_ = get_field_ordinal(meta, "density", stk::mesh::StateNP1);
+  dualNodalVolumeID_ = get_field_ordinal(meta, "dual_nodal_volume");
+  dpdxID_ = get_field_ordinal(meta, "dpdx");
+}
+
+void
+MomentumMassBDFNodeKernel::setup(Realm& realm)
+{
+  const auto& fieldMgr = realm.ngp_field_manager();
+
+  velocityNm1_ = fieldMgr.get_field<double>(velocityNm1ID_);
+  velocityN_ = fieldMgr.get_field<double>(velocityNID_);
+  velocityNp1_ = fieldMgr.get_field<double>(velocityNp1ID_);
+  densityNm1_ = fieldMgr.get_field<double>(densityNm1ID_);
+  densityN_ = fieldMgr.get_field<double>(densityNID_);
+  densityNp1_ = fieldMgr.get_field<double>(densityNp1ID_);
+  dualNodalVolume_ = fieldMgr.get_field<double>(dualNodalVolumeID_);
+  dpdx_ = fieldMgr.get_field<double>(dpdxID_);
+  dt_ = realm.get_time_step();
+  gamma1_ = realm.get_gamma1();
+  gamma2_ = realm.get_gamma2();
+  gamma3_ = realm.get_gamma3();
+}
+
+void
+MomentumMassBDFNodeKernel::execute(
+  NodeKernelTraits::LhsType& lhs,
+  NodeKernelTraits::RhsType& rhs,
+  const stk::mesh::FastMeshIndex& node)
+{
+  const int nDim = nDim_;
+
+  // deal with lumped mass matrix (diagonal matrix)
+  for ( int i = 0; i < nDim; ++i ) {
+    const NodeKernelTraits::DblType uNm1       = velocityNm1_.get(node, i);
+    const NodeKernelTraits::DblType uN         = velocityN_.get(node, i);
+    const NodeKernelTraits::DblType uNp1       = velocityNp1_.get(node, i);
+    const NodeKernelTraits::DblType rhoNm1     = densityNm1_.get(node, i);
+    const NodeKernelTraits::DblType rhoN       = densityN_.get(node, i);
+    const NodeKernelTraits::DblType rhoNp1     = densityNp1_.get(node, i);
+    const NodeKernelTraits::DblType dualVolume = dualNodalVolume_.get(node, i);
+    const NodeKernelTraits::DblType dpdx       = dpdx_.get(node, i);
+    const NodeKernelTraits::DblType lhsfac = gamma1_*rhoNp1*dualVolume/dt_;
+
+    rhs(i) += -(gamma1_*rhoNp1*uNp1 + gamma2_*rhoN*uN + gamma3_*rhoNm1*uNm1)*dualVolume/dt_ - dpdx*dualVolume;
+    const int row = i*nDim;
```

Comment:
`lhs` is a `3x3` matrix for `NodeKernel` this indexing is incorrect. Next line should be just `lhs(i, i)`.